apps/frontend/core-dapp/src/stores/entities/MetaStablePool.entity.ts

```diff
@@ -0,0 +1,123 @@
+    this.initTokens()
+    makeObservable(this, { tokens: observable })
+  }
+
+  initTokens(): void {
+    reaction(
+      () => this.poolTokens,
+      (poolTokens) => {
+        if (poolTokens === undefined) return
+        const tokens = poolTokens.tokens.map(
+          (address, index) => new MetaStablePoolToken(this, address, index)
+        )
+        runInAction(() => {
+          this.tokens = tokens
+        })
+      }
+    )
+  }
+
+  private getAmplificationParameter(
+    ...params: Parameters<GetAmplificationParameter>
+  ): ContractReturn<GetAmplificationParameter> {
+    return this.call<GetAmplificationParameter>('getAmplificationParameter', params)
+  }
+
+  private getPoolId(...params: Parameters<GetPoolId>): ContractReturn<GetPoolId> {
+    return this.call<GetPoolId>('getPoolId', params)
+  }
+
+  getPriceRateCache(...params: Parameters<GetPriceRateCache>): ContractReturn<GetPriceRateCache> {
+    return this.call<GetPriceRateCache>('getPriceRateCache', params)
+  }
+
+  private getSwapFeePercentage(
+    ...params: Parameters<GetSwapFeePercentage>
+  ): ContractReturn<GetSwapFeePercentage> {
+    return this.call<GetSwapFeePercentage>('getSwapFeePercentage', params)
+  }
+
+  get amplificationParameter():
+    | { value: BigNumber; isUpdating: boolean; precision: BigNumber }
+    | undefined {
+    const output = this.getAmplificationParameter()
+    if (output === undefined) return undefined
+    return {
+      value: output[0],
+      isUpdating: output[1],
+      precision: output[2],
+    }
+  }
+
+  get swapFeePercentage(): BigNumber | undefined {
+    return this.getSwapFeePercentage()?.[0]
+  }
+
+  get poolId(): string | undefined {
+    return this.getPoolId()?.[0]
+  }
+
+  get poolTokens():
+    | { tokens: string[]; balances: BigNumber[]; lastChangeBlock: BigNumber }
+    | undefined {
+    if (this.poolId === undefined) return undefined
+    const output = this.root.balancerVaultStore.getPoolTokens(this.poolId)
+    if (output === undefined) return undefined
+    return {
+      tokens: output[0],
+      balances: output[1],
+      lastChangeBlock: output[2],
+    }
+  }
+
+  // Balancer's calculation is based on data from their subgraph, which converts token with 0 price rate to 1
+  // and then apply the price rate to token balance in the reference below
+  // ref: https://github.com/balancer/balancer-sor/blob/john/v2-package-linear/src/pools/metaStablePool/metaStablePool.ts#L120
+  get allBalances(): BigNumber[] | undefined {
+    if (this.tokens === undefined) return undefined
+
+    const balances = this.tokens.map((token) => {
+      if (token.balance === undefined || token.priceRate === undefined) return undefined
+      const priceRate = token.priceRate.eq(0) ? parseEther('1') : token.priceRate
+      return token.balance.mul(priceRate).div(parseEther('1'))
+    })
+
+    if (balances.includes(undefined)) return undefined
+    return balances as BigNumber[]
```

Comment:
```suggestion
    const balances: BigNumber[] = []

    for (const token of this.tokens) {
      if (token.balance === undefined || token.priceRate === undefined) return undefined
      const priceRate = token.priceRate.eq(0) ? parseEther('1') : token.priceRate
      balances.push(token.balance.mul(priceRate).div(parseEther('1')))
    }

    return balances
```

Bu using `for..of` we can `return undefined` as soon as we encounter loading values, which is equivalent to what we're doing here. No need to cast `Array<BigNumbrer | undefined>` to `BigNumber[]` and more readable IMO.